### ISSUE TITLE
Any forward_pre_hook (except the first one self.register_forward_pre_hook(self._infer_parameters)) attached to LazyModuleMixin doesn't work right now 

### ISSUE THREAD

**Issue by ertrue** [pytorch/pytorch#60132]
## To Reproduce

1. Set up a module inherited from LazyModuleMixin
2. After module initialization, hook a function as hook_function(module, input)
3. run the forward function of the module.
4. Got the runtime error as "RuntimeError: OrderedDict mutated during iteration"

Example code to reproduce the runtime error:
```
import torch
from torch.nn.modules.lazy import LazyModuleMixin

class module_a(LazyModuleMixin, torch.nn.Module):
    def __init__(self):
        super().__init__()

    def initialize_parameters(self, input):
        return None

    def forward(self, input):
        return input

def hook_function(module, input):
    return input[0] + 1

A = module_a()
A.register_forward_pre_hook(hook_function)
output = A(torch.zeros(2, 2)) # Runtime Error
```

## Expected behavior

Extra function should be used as forward_pre_hook function beside _infer_parameters.

The previous example of code should be executed without error with expected output as:
```
tensor([[1., 1.],
        [1., 1.]])
```

## Environment

 - PyTorch version: 1.9.0.dev20210603+cpu
 - Is debug build: False
 - CUDA used to build PyTorch: None
 - ROCM used to build PyTorch: N/A

 - OS: CentOS Stream 8 (x86_64)
 - GCC version: (GCC) 8.4.1 20200928 (Red Hat 8.4.1-1)
 - Clang version: Could not collect
 - CMake version: Could not collect
 - Libc version: glibc-2.3.4

 - Python version: 3.6 (64-bit runtime)
 - Python platform: Linux-5.2.9-240_fbk16_hardened_4245_g53e65cc15fc6-x86_64-with-centos-8
 - Is CUDA available: False
 - CUDA runtime version: No CUDA
 - GPU models and configuration: No CUDA
 - Nvidia driver version: No CUDA
 - cuDNN version: No CUDA

## Additional context

The fix is on the module.py with one line change. Will submit a fix shortly.

cc @albanD @mruberry @jbschlosser

**Comment by ertrue** [comment]
A simple fix or workaround is in torch/nn/modules/module.py _call_impl() function, to change the iteration of hooks as list() of itertools.chain() as: 

From:
```
            for hook in itertools.chain(
                    _global_forward_pre_hooks.values(),
                    self._forward_pre_hooks.values()):
                result = hook(self, input)
                if result is not None:
                    if not isinstance(result, tuple):
                        result = (result,)
                    input = result
```

to
```
            for hook in list(itertools.chain(
                    _global_forward_pre_hooks.values(),
                    self._forward_pre_hooks.values())):
                result = hook(self, input)
                if result is not None:
                    if not isinstance(result, tuple):
                        result = (result,)
                    input = result
```

**Comment by jbschlosser** [comment]
Thanks for the report! Specifically, the error is triggered by the lazy module's initialization hook trying to remove itself from `_forward_pre_hooks` during hook iteration:
https://github.com/pytorch/pytorch/blob/85517a2b700a5abc0b38f53ce8c99404cd67db79/torch/nn/modules/lazy.py#L251

Proposed fix iterates over a copy of the hook list instead- I guess properly allowing the self-removing hook pattern is worth the efficiency hit? cc @albanD

**Comment by albanD** [comment]
Yes I think this is ok.
Especially because we have a "fast path" for people that don't use hooks, so this will only impact people that use them.

Note that you should update the other forward hook as well as they are most likely suffering from the same limitation.

The fix might not be to add a list btw but just remove the itertools.chain. As this function's goal is to avoid the copy that you want to do ;)

**Comment by jbschlosser** [comment]
Fixed in #60517.
